### PR TITLE
CDAP-12167 pipeline explicit transactions

### PR DESCRIPTION
**Pull request by albertshau**
Refactoring to make the initialize and destory methods in
ETLMapReduce, ETLSpark, and SmartWorkflow use explicit transaction
control. This is an improvement so that we don't run into
transaction timeouts unless a single plugin takes longer than the
short tx timeout. It also required in case alert publishing takes
longer than the short tx timeout.

Refactored the relevant plugin contexts to use the DatasetContext
provided by a TxRunnable instead of the MapReduceContext or
SparkClientContext as the DatasetContext. In order to do this,
added a context provider class that will return a new plugin
context given a DatasetContext.

### CODE REVIEW

**Comment by albertshau** [comment]
https://builds.cask.co/browse/CDAP-RUT1280

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java [hunk 1]
```diff
@@ -203,61 +210,144 @@ public void initialize() throws Exception {
...
+      if (submitterPlugin != null) {
+        submitterPlugin.prepareRun();
+        finishers.add(submitterPlugin);
       }
     }
-    finisher = finishers.build();
+
     hConf.set(SINK_OUTPUTS_KEY, GSON.toJson(sinkOutputs));
+    hConf.set(INPUT_ALIAS_KEY, GSON.toJson(inputAliasToStage));
+    finisher = finishers.build();
```
Comment (anwar6953) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java]: +1 for moving all the `finishers.add` to before the call to `finishers.builder`, even though calling the methods in the reverse order was also working...

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/submit/CompositeFinisher.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+   * @return builder for creating a composite finisher.
+   */
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  /**
+   * Builder to create a composite finisher
+   */
+  public static class Builder {
```
Comment (anwar6953) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/submit/CompositeFinisher.java]: What benefit is this builder implementation exposing? Seems like a `List<Finisher>` would be just as good.

Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/submit/CompositeFinisher.java]: true, will change it

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/StageSubmitterContext.java [hunk 3]
```diff
@@ -18,10 +18,13 @@
...
 import co.cask.cdap.api.messaging.MessagingAdmin;
 import co.cask.cdap.api.messaging.MessagingContext;
+import co.cask.cdap.etl.api.action.SettableArguments;
 
 /**
  * Similar to {@link TransformContext}, but also exposing functionality of {@link MessagingContext}.
  */
 public interface StageSubmitterContext extends MessagingContext, MessagingAdmin, TransformContext {
 
+  SettableArguments getArguments();
```
Comment (anwar6953) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/StageSubmitterContext.java:28]: This PR also resolves https://issues.cask.co/browse/CDAP-10974, right?

Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/StageSubmitterContext.java:28]: yeah, will update the commit message to mention that

**Review by anwar6953 — APPROVED** [review]

**Pull request merged**